core/src/main/scala/com/gu/tableversions/core/InMemoryTableVersions.scala

```diff
@@ -17,95 +13,63 @@ import com.gu.tableversions.core.util.RichRef._
-      message: UpdateMessage,
-      timestamp: Instant): F[Unit] =
-    allUpdates.update { prev =>
-      if (prev.contains(table)) prev
-      else
-        prev + (table -> List(
-          TableUpdate(
-            userId,
-            message,
-            timestamp = timestamp,
-            operations = List(InitTable(table, isSnapshot))
-          )))
+  override def commit(table: TableName, update: TableVersions.TableUpdate): F[Unit] = {
+    val applyUpdate: TableUpdates => Either[Exception, TableUpdates] = { currentTableUpdates =>
+      if (currentTableUpdates.contains(table)) {
+        val updated = currentTableUpdates + (table -> TableState(
+          currentVersion = update.header.id,
+          updates = currentTableUpdates(table).updates :+ update))
+        Right(updated)
+      } else
+        Left(unknownTableError(table))
     }
 
-  override def currentVersion(table: TableName): F[TableVersion] =
-    // Derive current version of a table by folding over the history of changes
-    for {
-      allTableUpdates <- allUpdates.get
-      tableUpdates <- allTableUpdates
-        .get(table)
-        .fold(F.raiseError[List[TableUpdate]](new Exception(s"Table '$table' not found")))(F.pure)
-      operations = tableUpdates.flatMap(_.operations)
-    } yield
-      if (isSnapShot(operations))
-        InMemoryTableVersions.latestSnapshotTableVersion(operations)
-      else
-        InMemoryTableVersions.applyPartitionUpdates(PartitionedTableVersion(Map.empty))(operations)
-
-  private def isSnapShot(operations: List[TableOperation]) = operations match {
-    case InitTable(_, isSnapshot) :: _ => isSnapshot
-    case _                             => throw new IllegalArgumentException("First operation should be InitTable")
+    allUpdates.modifyEither(applyUpdate).void
   }
 
-  override def commit(table: TableName, update: TableVersions.TableUpdate): F[TableVersions.CommitResult] = {
-
-    // Note: we're not checking invalid partition versions here as we suspect this problem will go
-    // away in a later iteration of this interface.
-
+  override def setCurrentVersion(table: TableName, id: CommitId): F[Unit] = {
     val applyUpdate: TableUpdates => Either[Exception, TableUpdates] = { currentTableUpdates =>
-      if (currentTableUpdates.contains(table)) {
-        val updated = currentTableUpdates + (table -> (currentTableUpdates(table) :+ update))
-        Right(updated)
-      } else
-        Left(new Exception(s"Unknown table '${table.fullyQualifiedName}'"))
+      currentTableUpdates.get(table) match {
+        case Some(currentTableState) =>
+          if (currentTableState.updates.exists(_.header.id == id)) {
+            val newTableState = currentTableState.copy(currentVersion = id)
+            val updated = currentTableUpdates + (table -> newTableState)
+            Right(updated)
+          } else
+            Left(unknownCommitId(id))
+
+        case None => Left(unknownTableError(table))
+      }
```

Comment:
Pattern matching instead of `fold`. Any specific reason?

Reply:
No particular reason - I'm just not in the habit of using `fold` on `Option`s. But I can change...

Reply:
BTW do you also prefer `fold` to `map` + `getOrElse`?

Reply:
I don't mind to be honest, as long as it is consistant throughout the codebase

Reply:
I think `getOrElse` would usually attract the ire of fp fanatics ;-)

Reply:
Why? `Option.get` is obviously bad, but what's the problem with `getOrElse`?

Reply:
Unlike `fold` it doesn't force the return type to be the same as the `Option`.
```
scala> Some(1).getOrElse("something")
res0: Any = 1

scala> Some(1).fold("something")(identity)
<console>:12: error: type mismatch;
 found   : Int
 required: String
       Some(1).fold("something")(identity)
```

Reply:
That's all the justification I need, `fold` it is then!

Reply:
As mentioned above, this doesn't work because it doesn't infer the right type:

```scala
Some(1).fold(Left("Empty!"))(n => Right(n + 2))
```

So you have to explicitly provide the type:

```scala
Some(1).fold[Either[String, Int]](Left("Empty!"))(n => Right(n + 2))
```

Unless I'm missing something? (I'm still going with the `fold` anyway)

Reply:
I don't think you are missing anything. Type inference in Scala is quite limited